include/inviwo/core/util/factory.h

```diff
@@ -64,55 +65,100 @@ class FactoryObservable : public Observable<FactoryObserver<T>> {
 /**
  * A common factory base class
  */
-class IVW_CORE_API FactoryBase {
+template <typename K = std::string_view>
+class FactoryBase {
 public:
     FactoryBase() = default;
     virtual ~FactoryBase() = default;
     FactoryBase(const FactoryBase&) = delete;
     FactoryBase& operator=(const FactoryBase&) = delete;
     FactoryBase(FactoryBase&&) = default;
     FactoryBase& operator=(FactoryBase&&) = default;
+
+    virtual bool hasKey(K key) const = 0;
 };
+extern template class IVW_CORE_TMPL_EXP FactoryBase<std::string_view>;
 
 /**
  * An abstract factory interface. Inherits virtually from factory base, since an implementation
  * might implement several factory interfaces
  * T Models the object created, T will be constructed using Args...
- * K Models a key used to look up T
  */
-template <typename T, typename K = const std::string&, typename... Args>
-class Factory : public virtual FactoryBase {
+template <typename T, typename K = std::string_view, typename... Args>
+class Factory : public virtual FactoryBase<K> {
 public:
     Factory() = default;
     virtual std::unique_ptr<T> create(K key, Args... args) const = 0;
     virtual bool hasKey(K key) const = 0;
 };
 
+template <typename M, typename Key, typename K>
+class FactoryRegister : public FactoryObservable<M> {
+public:
+    // The factory will not assume ownership over obj, although is assumes that obj will be
+    // valid for the lifetime of the factory
+    virtual bool registerObject(M* obj);
+    virtual bool unRegisterObject(M* obj);
+
+    bool hasKey(K key) const;
+    std::vector<Key> getKeys() const;
+
+protected:
+    std::map<Key, M*, std::less<>> map_;
+};
+
+template <typename K>
+struct FactoryKeyType {
+    using type = std::remove_cv_t<std::remove_reference_t<K>>;
+};
+template <>
+struct FactoryKeyType<std::string_view> {
+    using type = std::string;
+};
+template <typename K>
+using factoryKeyType = typename FactoryKeyType<K>::type;
+
 /**
  * T Models the object created
  * M Models a object with a function create(K key, Args...) that can create objects of type T with
  * constructor T(Args...)
  * M would usually be a "factory object" type
  * K Models a key used to look up T
  */
-template <typename T, typename M, typename K = const std::string&, typename... Args>
-class StandardFactory : public Factory<T, K, Args...>, public FactoryObservable<M> {
+template <typename T, typename M, typename K = std::string_view, typename... Args>
+class StandardFactory : public Factory<T, K, Args...>,
+                        public FactoryRegister<M, factoryKeyType<K>, K> {
```

Comment:
Not sure it there is a benefit of doing the using declaration for FactoryKeyType<K>::type 
Less readable.

Later, we use StandardFactory like this:
`StandardFactory<T, M, string_view,...>`
then 
```
FactoryRegister::hasKey(string_view)
std::vector<string> FactoryRegister::getKeys()
```

Would it even work with `StandardFactory<T, M, string,...>`? If not, are all of these templates even needed?

Reply:
Would rename K to Key while at it :)

Reply:
`factoryKeyType<K> -> typename FactoryKeyType<K>::type`
which almost twice as long...
Not sure what you mean by the other comments. Yes the templates are needed?

Reply:
Ah, I missed the typename part

Reply:
Nice commit getting rid of the hasKey type. Can we now also assume
`vector<string> getKeys()`
:)

Reply:
the FactoryRegister base class this has a `std::vector<Key> getKeys()`

Reply:
factoryKeyType<K> is still there.